source/slang/slang-check-impl.h

```diff
@@ -28,22 +28,51 @@ namespace Slang
 
     // A flat representation of basic types (scalars, vectors and matrices)
     // that can be used as lookup key in caches
-    enum class BasicTypeKey : uint16_t
+    struct BasicTypeKey
     {
-        Invalid = 0xffff,                 ///< Value that can never be a valid type
+        uint32_t baseType : 8;
+        uint32_t dim1 : 4;
+        uint32_t dim2 : 4;
+        uint32_t knownConstantBitCount : 8;
+        uint32_t knownNegative : 1;
+        uint32_t reserved : 7;
+        uint32_t getRaw() const
+        {
+            uint32_t val;
+            memcpy(&val, this, sizeof(uint32_t));
+            return val;
+        }
+        bool operator==(BasicTypeKey other) const
+        {
+            return getRaw() == other.getRaw();
+        }
+        static BasicTypeKey invalid() { return BasicTypeKey{ 0xff, 0, 0 }; }
     };
 
     SLANG_FORCE_INLINE BasicTypeKey makeBasicTypeKey(BaseType baseType, IntegerLiteralValue dim1 = 0, IntegerLiteralValue dim2 = 0)
     {
         SLANG_ASSERT(dim1 >= 0 && dim2 >= 0);
-        return BasicTypeKey((uint8_t(baseType) << 8) | (uint8_t(dim1) << 4) | uint8_t(dim2));
+        return BasicTypeKey{ uint8_t(baseType), uint8_t(dim1), uint8_t(dim2) };
     }
 
-    inline BasicTypeKey makeBasicTypeKey(Type* typeIn)
+    inline BasicTypeKey makeBasicTypeKey(Type* typeIn, Expr* exprIn = nullptr)
     {
         if (auto basicType = as<BasicExpressionType>(typeIn))
         {
-            return makeBasicTypeKey(basicType->baseType);
+            auto rs = makeBasicTypeKey(basicType->baseType);
+            if (auto constInt = as<IntegerLiteralExpr>(exprIn))
+            {
+                auto value = constInt->value;
+                if (value < 0)
+                {
+                    rs.knownNegative = 1;
+                    value = -value;
+                }
+                rs.knownConstantBitCount = 1;
+                while (value >>= 1)
```

Comment:
... and then sign extension will cause this to not terminate.

Reply:
Fixed.